### ISSUE TITLE
make repl_history.jl less racy

### ISSUE THREAD

**Issue by StefanKarpinski** [JuliaLang/julia#37015]
We're [hearing](https://github.com/JuliaLang/julia/issues/36895) from users on clusters with shared home directories that their `~/.julia/logs/repl_history.jl` files are frequently corrupted. It would be good to make this less racy, either by doing some file locking with Pidfile or by some other mechanism.

**Comment by stevengj** [comment]
Just use `git` to manage the history.  Or maybe PostgreSQL?

**Comment by StefanKarpinski** [comment]
Requiring a database or git (which we're trying to reduce and eventually eliminate dependencies on) for REPL history seems a bit excessive. How would one use git for this?

**Comment by stevengj** [comment]
Sorry, I thought it was obvious that I was joking. 🤷

**Comment by StefanKarpinski** [comment]
That went _right_ over my head 😬

**Comment by notinaboat** [comment]
I have this problem a few times per day.

Other's have the same problem with clusters: https://github.com/JuliaLang/julia/issues/36895#issuecomment-672953081 

The thing* I'm working on has two Julia processes on my intel Mac and another three each on a seperate Raspberry Pi.
All the processes share a JULIA_DEPOT_PATH. The Raspberry Pis have this directory mounted via NFS.
(sometimes there is also a qemu vm in the mix)

I can't see file locking ever solving this problem. Even if there was a reliable cross-platform (and cross-filesystem?) way to do the locking, it wouldn't be great to have everyone waiting for one node in a cluster to release it's file lock.

Perhaps it would be simpler for every process to log to `logs/repl_history.d/$host.$pid`.
Then, when loading the history, iterate through all of `logs/repl_history.d/*` in parallel, pushing items into the REPLHistoryProvider in timestamp order.

I imaging an eventual consistency scheme for coalescing files where at history loading time:
 - any file loaded that was last modified more than one day ago is deleted
 - content from the delete file is saved under the new `$host.$pid` file of the process doing the history loading.

This could result in duplication if who processes both decided to do deletion at the same time, but exact duplicates of timestamp and command could simply be ignored at load time.

@StefanKarpinski, if you would support a PR along these lines I would consider working on it. I can't quite bring myself to use a startup.jl hack like this: https://github.com/JuliaLang/julia/issues/36895#issuecomment-681072278

<hr>
* the "thing" is a test harness for an embedded industrial control system. The R'Pis simulate sensor inputs and record outputs from the control board. The Mac does test scenario orchestration, log recording and log analysis. Each Julia process has a REPL accessible via tmux for interactive use. 

https://github.com/JuliaLang/julia/blob/7a6336d41a4064c0bf3649fb3b1a2ec3531a7dd8/stdlib/REPL/src/REPL.jl#L621-L626

**Comment by Moelf** [comment]
Do we know how does bash/zsh/fish/ipython deal with this?

**Comment by caleb-allen** [comment]
I appreciate the approach taken by fish shell documented on [this issue](https://github.com/fish-shell/fish-shell/issues/825)

`fish` doesn't implement any automatic synchronization mechanism because of the added performance penalties and complexity, and instead have a [`history merge` command](https://fishshell.com/docs/current/cmds/history.html):

> Immediately incorporates history changes from other sessions. Ordinarily fish ignores history changes from sessions started after the current one. This command applies those changes immediately.

It strikes me as a nice balance. It's a simple command which exposes the feature for those that need it, but it avoids adding complexity that would come with an automatic feature.
